url8d3c4d17f5_catseye_JaC64_tgz-pJ8-JaC64J8/src/com/dreamfabric/jac64/C1541.java

```diff
@@ -0,0 +1,532 @@
+ * emulation.
+ *
+ * Created: Tue Mar 28 22:06:31 2006
+ *
+ * @author Joakim Eriksson, Dreamfabric.com
+ * joakime@sics.se
+ * @version 1.0
+ */
+public class C1541 extends Observable {
+
+  public static final int IO_OFFSET = CPU.IO_OFFSET;
+
+  public static final boolean DEBUG = true; //false;
+
+  public static final int SERIAL_ATN = (1 << 3);
+  public static final int SERIAL_CLK_OUT = (1 << 4);
+  public static final int SERIAL_DATA_OUT = (1 << 5);
+  public static final int SERIAL_CLK_IN = (1 << 6);
+  public static final int SERIAL_DATA_IN = (1 << 7);
+
+  public static final int TALK = 0x40;
+  public static final int LISTEN = 0x20;
+  public static final int DATA = 0x60;
+  public static final int OPEN = 0xf0;
+  public static final int CLOSE = 0xe0;
+
+  // Cpu memory
+  private int[] memory;
+  // Disk (.d64) reader
+  private C64Reader reader;
+
+  private static final int IDLE = 0;
+  private static final int ATN = 1;
+  private static final int RECEIVING = 2;
+  private static final int SENDING = 3;
+  private static final int READ_BIT = 4;
+  private static final int WAIT_BIT = 5;
+  private static final int READ_BYTE = 6;
+  private static final int WRITE_BYTE = 7;
+
+  private static final int LOAD_FILE = 1;
+  private static final int SAVE_FILE = 2;
+  private static final int LOGICAL_CHANNEL = 3;
+
+  private static final int ATN_SEEN = 10;
+  private static final int ATN_READ_BIT = 11;
+  private static final int ATN_WAIT_BIT = 12;
+
+  private static final int WAIT_LISTENER_READY = 13;
+  private static final int WRITE_BIT_CLK1 = 14;
+  private static final int WRITE_BIT_CLK2 = 15;
+  private static final int WRITE_END = 16;
+  private static final int WAIT_LISTENER_EOI_HANDSHAKE = 17;
+
+  private static final int READ_FILENAME = 1;
+
+  private DiskChannel channel[] = new DiskChannel[16];
+
+  private boolean atnLast = false;
+  private int mode = IDLE;
+  private long eoiTimeout = 0;
+  private int eoi;
+  private boolean lastChar = false;
+
+  private int role = 0;
+  private int floppyMode = 0;
+  private int floppyChannel = 0;
+  private String filename;
+
+  // For example when reading bytes and waits for EOI
+  private long waitTimeout = 0;
+
+  private int readMode = READ_FILENAME;
+  private String tmpFilename = "";
+  /**
+   * Creates a new <code>C1541</code> instance.
+   *
+   */
+  public C1541(int[] memory) {
+    this.memory = memory;
+    reader = new C64Reader();
+    for (int i = 0, n = 16; i < n; i++) {
+      channel[i] = new DiskChannel(i);
+    }
+  }
+
+  public void reset() {
+    mode = IDLE;
+    role = 0;
+    rbState = 0;
+    rbByte = 0;
+    floppyMode = 0;
+    floppyChannel = 0;
+    clockHi();
+  }
+
+  public C64Reader getReader() {
+    return reader;
+  }
+
+  public void tick(long cycles) {
+    if (waitTimeout != 0 && waitTimeout < cycles) {
+      System.out.print(".");
+      tick(cycles, true);
+    }
+  }
+
+  public void tick(long cycles, boolean timeout) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 20. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.